tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -1963,6 +1963,47 @@ bool DeclResultIdMapper::createStageVars(
     // function/parameter/variable. All are DeclaratorDecls.
     stageVarInstructions[cast<DeclaratorDecl>(decl)] = varInstr;
 
+    // Special case: The DX12 SV_InstanceID always counts from 0, even if the
+    // StartInstanceLocation parameter is non-zero. gl_InstanceIndex, however,
+    // starts from firstInstance. Thus it doesn't emulate actual DX12 shader
+    // behavior. To make it equivalent, SPIR-V codegen should emit:
+    // SV_InstanceID = gl_InstanceIndex - gl_BaseInstance
+    // Unfortunately, this means that there is no 1-to-1 mapping of the HLSL
+    // semantic to the SPIR-V builtin. As a result, we have to manually create
+    // a second stage variable for this specific case.
```

Comment:
This comment looks good, but I am curious there is a document about this rule.
If it exists, please add the link here.

Reply:
The exact meanings of these builtins are documented here: [https://www.khronos.org/registry/vulkan/specs/1.0-wsi_extensions/html/vkspec.html#interfaces-builtin-variables](url)

"InstanceIndex begins at the firstInstance parameter to vkCmdDraw or vkCmdDrawIndexed or at the firstInstance member of a structure consumed by vkCmdDrawIndirect or vkCmdDrawIndexedIndirect."

"BaseInstance is the firstInstance parameter to a direct drawing command or the firstInstance member of a structure consumed by an indirect drawing command."

HLSL semantics are described here: [https://docs.microsoft.com/en-us/windows/win32/direct3d11/d3d10-graphics-programming-guide-input-assembler-stage-using](url):

"It is a 32-bit unsigned integer whose default value is 0."

There is no mention of it starting from anything else than 0.